### PR TITLE
Extension Metadata

### PR DESCRIPTION
**Pull request by carlopi**
### Why?
* Observability of what extension you have at the moment on your system
* Allow checks on consistency of both duckdb's version and platform before actually loading a given binary

Both together allows to avoid a class of bugs where if for any reason duckdb was instructed to load a given extension (say via the explicit LOAD 'path/to/file'), and there was a platform mismatch this lead to weird and complex to debug crashes.

Also, observability of current available extensions allow more precise informations while tracking for bugs related to extensions.

### How
Extensions binary now have a 512 bytes payload added at the end. Of those 256 are reserved for metadata and 256 for the signature.

Of the 256 metadata bytes only 4 slots of 32 bytes are actually used, with 128 bytes reserved for future usage.
The current 4 slots are:
* A slot that defaults to "4". This serve as the metadata version, and currently doubles as the number of available slots, but meaning can and will change
* A slot for the duckdb_platform the extension has been build for
* A slot for the duckdb_version the extension has been build for
* A slot for the extension_version of the extension. Currently can be either
  - a string containing a git reference (either tag or hash) for out of tree extensions
  - a string containing the duckdb_version for in-tree extensions
  - a custom string if build_loadable_extension_directory is invoked directly, see test/extension/CMakeLists.txt where for loadable_extension_demo is set to "default-version"

Given duckdb-wasm extensions need to have the payload be part of a [custom section](https://webassembly.github.io/spec/core/appendix/custom.html), a set of magic bytes are added before the 512 bytes payload, and those allows the script to be equivalent between wasm & native cases, and has the property that given in input a valid WebAssembly module a valid WebAssembly module will be returned in output. Custom section specification encodes basically the length of the segment that will follow.

NOTE: Signing script has also been changed, now any extension build by duckdb will have an empty signature, so when extension are signed first file need to be truncated and only then signature can be added.

### Demo
If you don't want to wait for CI to try this out, steps are like:

1. Add those lines at the bottom of `extension/extension_config_local.cmake`
```
duckdb_extension_load(json)
duckdb_extension_load(autocomplete)
duckdb_extension_load(icu DONT_LINK)
duckdb_extension_load(httpfs DONT_LINK)
duckdb_extension_load(tpch DONT_LINK)
duckdb_extension_load(tpcds DONT_LINK)
duckdb_extension_load(sqlite_scanner
        DONT_LINK LOAD_TESTS
        GIT_URL https://github.com/duckdb/sqlite_scanner
        GIT_TAG 70a4c411434290b4af704940987a324d84af0024
        APPLY_PATCHES
        )
```
Then run:
```bash
make clean && ENABLE_EXTENSION_AUTOLOADING=1 ENABLE_EXTENSION_AUTOINSTALL=1  GEN=ninja make
```
building DuckDB CLI in a setting similar to CLI.
```bash
./build/release/duckdb -unsigned -cmd "SET custom_extension_repository='build/release/repository';"
```
then executing:
```sql
CREATE TABLE services AS FROM 'https://blobs.duckdb.org/stations.csv.gz'; --- autoinstall+load httpfs
ATTACH 'hi.db' (TYPE SQLITE);  --- autoinstall+load sqlite
SELECT * EXCLUDE(description, install_path, aliases) FROM duckdb_extensions();
```
results in:
```
┌──────────────────┬─────────┬───────────┬──────────────────────────────────┐
│  extension_name  │ loaded  │ installed │        extension_version         │
│     varchar      │ boolean │  boolean  │             varchar              │
├──────────────────┼─────────┼───────────┼──────────────────────────────────┤
│ arrow            │ false   │ false     │                                  │
│ autocomplete     │ true    │ true      │                                  │
│ aws              │ false   │ false     │                                  │
│ azure            │ false   │ false     │                                  │
│ excel            │ false   │ false     │                                  │
│ fts              │ false   │ false     │                                  │
│ httpfs           │ true    │ true      │ 38047b7c8d                       │
│ iceberg          │ false   │ false     │                                  │
│ icu              │ false   │ false     │                                  │
│ inet             │ false   │ false     │                                  │
│ jemalloc         │ false   │ false     │                                  │
│ json             │ true    │ true      │                                  │
│ motherduck       │ false   │ false     │                                  │
│ mysql_scanner    │ false   │ false     │                                  │
│ parquet          │ true    │ true      │                                  │
│ postgres_scanner │ false   │ false     │                                  │
│ shell            │ true    │           │                                  │
│ spatial          │ false   │ false     │                                  │
│ sqlite_scanner   │ true    │ true      │ 70a4c411434290b4af704940987a324d │
│ substrait        │ false   │ false     │                                  │
│ tpcds            │ false   │ false     │                                  │
│ tpch             │ false   │ false     │                                  │
├──────────────────┴─────────┴───────────┴──────────────────────────────────┤
│ 22 rows                                                         4 columns │
└───────────────────────────────────────────────────────────────────────────┘
```

----

Up to here the same behaviour would be true also with main, but the extension_version column would be missing, now the interesting part, what's improved by this PR:

#### Error caught by this PR:
* platform mismatch (currently potential for crashes)
* version mismatch (currently happens after dlopen, that is too late for recovery)
* allow more informative error messages:

Say for example we do a non-sensical `LOAD 'README.md';`:
1) duckdb main, default settings:
```
IO Error: Extension "README.md" could not be loaded because its signature is either missing or invalid and unsigned extensions are disabled by configuration.
Start the shell with the -unsigned parameter to allow this (e.g. duckdb -unsigned).
```
2) duckdb main, with `-unsigned`:
```
IO Error: Extension "README.md" could not be loaded: dlopen(README.md, 0x0006): tried: 'README.md' (not a mach-o file), ...
```
3) after this PR, default:
```
IO Error: Extension "README.md" could not be loaded because its signature is either missing or invalid and unsigned extensions are disabled by configuration.
Start the shell with the -unsigned parameter to allow this (e.g. duckdb -unsigned).
Extension "README.md" do not have metadata compatible with DuckDB loading it (version bc432187d4, platform osx_arm64)
```
4) after this PR, with `-unsigned`:
```
Invalid Input Error: Extension "README.md" do not have metadata compatible with DuckDB loading it (version bc432187d4, platform osx_arm64)
```

There have been a hosts of issue opened over time that were connected to some from of misuse of extension or extension repository that gave unclear error messages that can now be improved.

Also this allows check to happen BEFORE linking-in a library that (even if signed) might not be compatible (say in case of platform mismatch).

----

### Changes required for extension developers
* extension_version function can be removed/deprecated from extensions as not used anymore
* extensions will need to use the new script to be updated, after this PR I will need to change duckdb's extensions CI to rely on new metadata generation & signing script
* `FROM duckdb_extensions();` now returns an additional column, the extension_version. If somehow relies on exact schema, that will need to be adapted.

### Up next
Currently extension_version is correctly propagated only for loaded (non-built-in) extensions, a similar logic can be also applied at INSTALL time doing cursory checks on platform version / extension version and allowing to solve some issues also on INSTALL side.
Also on top of this UPDATE logic becomes viable.

### Problems I am aware of
I am aware of one problem with the current situation that is been brought to light by this PR: if the _init function calls Load<> explicitly on itself, the extension_version field will end up being empty.
This is for example the case of the TPCH extension in its current form.
Basically add `FORCE INSTALL tpch; LOAD tpch;` to the previous example, it would lead to an empty `extension_version` shown for tpch.

Solution here is likely removing the recursive call from tpch_extension_init, but given `extension_version` is at the moment an optional field it can be moved to an independent PR and re-assessed. Another solution is splitting up `db.SetExtensionLoaded(extension, res.extension_version);`, setting extensions as loaded BEFORE calling init but calling callbacks AFTER init. This can be solved independently, not a blocker for this PR.

### CODE REVIEW

**Comment by carlopi** [comment]
This is a rework / refactor of https://github.com/duckdb/duckdb/pull/10460.

Thanks to @samansmink for the various feedback on this.

> I think the ExtensionHelper::TryInitialLoad has reached a point here where it is a bit too long to be readable, could we split this one up in some smaller functions?

This comment of his (https://github.com/duckdb/duckdb/pull/10460#discussion_r1516238226) has not been addressed, but will be done while absorbing the patches for duckdb wasm.

**Comment by carlopi** [comment]
More testing and cleanup.

Now extension_version metadata will be determined, in order, by:
* eventual EXTENSION_VERSION string parameter provided to duckdb_extension_load
* eventual GIT_TAG string parameter provided to duckdb_extension_load
* if the SOURCE_DIR is provided AND git appears to be functional, use the output of "cd $SOURCE_DIR && git log -1 --format=%h"
* else, empty string, ""

File: src/function/table/system/duckdb_extensions.cpp [hunk 1]
```diff
@@ -98,16 +102,20 @@ unique_ptr<GlobalTableFunctionState> DuckDBExtensionsInit(ClientContext &context
 	});
 #endif
 	// now check the list of currently loaded extensions
-	auto &loaded_extensions = db.LoadedExtensions();
-	for (auto &ext_name : loaded_extensions) {
+	auto &loaded_extensions = db.LoadedExtensionsData();
```
[the remark this answers is not in the document]
Comment (carlopi) [reply src/function/table/system/duckdb_extensions.cpp:105]: I agree this would be appropriate, but I would feel more comfortable doing that in a separate PR.

I would say goal of this PR is:
1. introducing metadata
2. adding validation on load (that doubles as a check that metadata has to be right)
3. introducing extension_version metadata (when loaded, shown also via `FROM duckdb_extensions()`)

Next one would be for:
1. reading metadata from the local extension_repository, shown also via `FROM duckdb_extensions()`
2. some form of validations on INSTALL (platform + version, or bypassed when allow_extensions_metadata_mismatch is set && unsigned extensions)

One left over question, connected to this: how to propagate version for statically built in extensions? Is that ever relevant?

File: CMakeLists.txt [hunk 2]
```diff
@@ -820,20 +816,34 @@ function(build_loadable_extension_directory NAME OUTPUT_DIRECTORY PARAMETERS)
       COMMAND emcc $<TARGET_FILE:${TARGET_NAME}>.lib -o $<TARGET_FILE:${TARGET_NAME}> -sSIDE_MODULE=1 -O3
       )
   endif()
+  add_custom_command(
+    TARGET ${TARGET_NAME}
+    POST_BUILD
+    COMMAND
+    printf "${DUCKDB_NORMALIZED_VERSION}" > duckdb_version_file
+    COMMAND
+    printf "${EXTENSION_VERSION}" > extension_version_file || ( echo "CMake appears not to propagate correctly version information for the extension, using empty string" && rm extension_version_file && touch extension_version_file )
```
Comment (samansmink) [inline CMakeLists.txt:825]: Just wondering now, does this work on windows?

File: scripts/append_metadata.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
```
Comment (samansmink) [inline scripts/append_metadata.sh:1]: Again here, I have some worries about portability of this script. I guess on CI this may be fine, but if this breaks most people's windows build that would be a problem

**Review by samansmink — COMMENTED** [review]
left 2 comments (discussed in person already)

File: CMakeLists.txt (hunk 2 above)
Comment (carlopi) [reply CMakeLists.txt]: This was not working on windows, moved to echo_append

File: scripts/append_metadata.sh (hunk 3 above)
Comment (carlopi) [reply scripts/append_metadata.sh]: I went back to... invoking a cmake script, see `scripts/append_metadata.cmake`.

This should properly solve portability.

**Comment by carlopi** [comment]
Failures are related to azure, so blocked on https://github.com/duckdb/duckdb/pull/11595.

Yesterday afternoon (when CI was still functional) this was green (but for nightly.yml) on my fork that runs more tests (e.g.: https://github.com/carlopi/duckdb/actions/runs/8617487833)

File: scripts/append_metadata.cmake [hunk 4]
```diff
@@ -0,0 +1,68 @@
+cmake_minimum_required(VERSION 3.5)
+
+# Usage: cmake -DEXTENSION=path/to/extension.duckdb_extension -DPLATFORM_FILE=README.md -DDUCKDB_VERSION=tag1 -DEXTENSION_VERSION=tag2 -P scripts/append_metadata.cmake
+# Currently hardcoded to host up to 8 fields
+# Example: ./scripts/append_metadata.sh file.duckdb_extension git_hash_duckdb_file git_hash_extension_file platfrom_file
```
Comment (samansmink) [inline scripts/append_metadata.cmake:5]: nit: this still say append_metadata.sh

Also with CMake's very weird partial support for null bytes (requiring this horrible `null.txt` hack), i would like to have some double checks here that our header is written correctly. Maybe we can add a check afterwards that confirms the header is correct. 

I will add this in a follow-up PR to not block this one

**Review by samansmink — APPROVED** [review]

File: scripts/append_metadata.cmake (hunk 4 above)
Comment (carlopi) [reply scripts/append_metadata.cmake:5]: Will fix those in a follow up.

**Comment by carlopi** [comment]
This has been tested in the CI setting in all various configurations (in particular on my fork, where more workflows are enabled: https://github.com/carlopi/duckdb/actions?query=branch%3Aextension_metadata++), tested locally on MacOS and Windows.

Weaker point was metadata generation / appending, now that's done via invoking a CMake script, that should ensure portability.

File: src/main/extension/extension_load.cpp [hunk 5]
```diff
@@ -119,15 +152,66 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileSystem &fs, const str
...
+
+	const std::string engine_version = std::string(GetVersionDirectoryName());
+	const std::string engine_platform = std::string(DuckDB::Platform());
+
+	auto handle = fs.OpenFile(filename, FileFlags::FILE_FLAGS_READ);
+
+	idx_t file_size = handle->GetFileSize();
+
+	if (file_size < 1024) {
+		throw InvalidInputException(
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:167]: "File \"%s\" is not a DuckDB extension. Valid DuckDB extensions must be at least 1024 bytes.

File: src/main/extension/extension_load.cpp [hunk 6]
```diff
@@ -119,15 +152,66 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileSystem &fs, const str
...
+	std::string extension_version = FilterZeroAtEnd(metadata_field[3]);
+
+	string metadata_mismatch_error = "";
+	{
+		char a[32] = {0};
+		a[0] = '4';
+		if (strncmp(a, metadata_field[0].data(), 32) != 0) {
+			// metadata do not looks right, add this to the error message
+			metadata_mismatch_error =
+			    "\n" + StringUtil::Format("Extension \"%s\" do not have metadata compatible with DuckDB "
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:195]: File \"%s\" is not a DuckDB extension. The metadata at the end of the file is invalid. This version of DuckDB can only load extensions that are compiled for DuckDB version %s, platform %s.

File: src/main/extension/extension_load.cpp [hunk 7]
```diff
@@ -119,15 +152,66 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileSystem &fs, const str
...
+			                              "loading it (version %s, platform %s)",
+			                              filename, engine_version, engine_platform);
+		} else if (engine_version != extension_duckdb_version || engine_platform != extension_duckdb_platform) {
+			metadata_mismatch_error = "\n" + StringUtil::Format("Extension \"%s\" (version %s, platfrom %s) does not "
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:199]: File \"%s\" is a DuckDB extension, but it was built for a different DuckDB version. The file was built for DuckDB version \"%s\", but we can only load extensions built for DuckDB version \"%s\".

File: src/main/extension/extension_load.cpp [hunk 8]
```diff
@@ -119,15 +152,66 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileSystem &fs, const str
...
+			                                                    "match DuckDB loading it (version %s, platform %s)",
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:200]: File \"%s\" is a DuckDB extension, but it was built for a different platform. The file was built for the platform \"%s\", but we can only load extensions built for platform \"%s \".

File: src/main/extension/extension_load.cpp [hunk 9]
```diff
@@ -119,15 +152,66 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileSystem &fs, const str
...
+	metadata_segment.resize(512);
+
+	const std::string engine_version = std::string(GetVersionDirectoryName());
+	const std::string engine_platform = std::string(DuckDB::Platform());
+
+	auto handle = fs.OpenFile(filename, FileFlags::FILE_FLAGS_READ);
+
+	idx_t file_size = handle->GetFileSize();
+
+	if (file_size < 1024) {
```
Comment (Mytherin) [inline src/main/extension/extension_load.cpp:166]: Should this be 512? And maybe move this to a `constexpr` instead of a magic number

File: test/api/test_reset.cpp [hunk 10]
```diff
@@ -82,6 +82,7 @@ OptionValueSet &GetValueForOption(const string &name) {
 	    {"file_search_path", {"test"}},
 	    {"force_compression", {"uncompressed", "Uncompressed"}},
 	    {"home_directory", {"test"}},
+	    {"allow_extensions_metadata_mismatch", {"true"}},
```
Comment (Mytherin) [inline test/api/test_reset.cpp:85]: This should only work for `-unsigned`

**Review by Mytherin — COMMENTED** [review]
Thanks! Some more comments for @samansmink

**Comment by softprops** [comment]
👋 I just upgraded an extension built with zig was working in 0.9.2 that stopped working after upgrading to 0.10.2, I think because of this change, discussion [here](https://github.com/duckdb/duckdb/discussions/12000).

For those not building extensions with cmake is there a documentation in terms of a formal specification of what is needed outside of this pull description? 

I'd love to not have to add cmake to my toolchain and I think it should be possible to derive what's needed as along as its formally specified somewhere.

**Comment by samansmink** [comment]
Hey @softprops! Currently our extension build is CMake based. We are still actively developing on this, and we don't have a formal spec right now. As the build stabilises we will consider publishing a formal spec for this so that other build systems can be more easily created/maintained.

Happy to take answer any questions regarding to how duckdb extensions are built though, so feel free to ping us through GitHub or the DuckDB discord!

**Comment by softprops** [comment]
@samansmink no worries. I got this sorted out in a zig build script here https://github.com/duckdb/duckdb/discussions/12000#discussioncomment-9400841
